dotnet/Nerdbank.GitVersioning
# Caching git heights

## Summary
Adds caching of git heights. As calculating the height of repositories with large volumes of commits is expensive, caching the git heights can save time in the following circumstances:
- Repetitive invocations of the `GetBuildVersion` msbuild task- e.g. during `dotnet pack`'ing a project, the `GetBuildVersion` task is invoked four times
- Incremental versioning- in cases where a cached height is available for older commits, this value will be used avoid the cost of recalculating the entire git height (only new commits will need to be traversed)

### Implementation details
### When can cached results be used?
- The base version to calculate the height for is the same as the cached base version
- The relative path of the cached height matches the requested relative path 
- The commit to calculate the height for is either the same OR a child of the cached commit id

### Opting out
The caching behavior is enabled by default but can be opted-out by setting the new `NerdbankGitVersioningUseHeightCache` msbuild property to `false`

### Cache file
A `version.cache.json` file is created per-project with contents that look like this:
```
/*Cached commit height, created by Nerdbank.GitVersioning. Do not modify.*/{"BaseVersion":"1.0","Height":1636,"CommitId":"f30bb9e26af1dc80f5c38f43d9d6a7b4f770bb14","RelativeProjectDir":""}
```

## Testing
### Automated testing
- Verified `GitHeightCache` can serialize + deserialize heights correctly
- Verified caching has a measurable impact on performance for cases when there are many commits to traverse
- All existing tests pass

### Manual testing
Consumed locally-packed version of `Nerdbank.GitVersioning` in a C# project and added ~1500 commits, verifying:
- Height caching takes effect on second build, dramatically decreasing build time (~10s -> ~1s)
- Adding an additional single commit can leverage the cached version for the previous ~1500 commits but adds the latest
- Setting the `NerdbankGitVersioningUseHeightCache` property to `false` bypasses the caching behavior.
